Feature/maintenance notifications

kevinglover opened the pull request.

kevinglover committed:
Merge branch 'release/0.5' into develop

kevinglover committed:
Updated dev version number

kevinglover committed:
Fixed typo in CSS

kevinglover committed:
Integrate MX banners and Downtime

- Added maintenance banners
- Added django-downtime

kevinglover committed:
Update Manifest

kevinglover committed:
DowntimePeriod model and Updated test

- DowntimePeriod extends downtime.periods model
- Title more meaningful
- Updated tests

kevinglover committed:
Added management commands for ending downtime

- Customized commands needed until bug in django-downtime is fixed

kevinglover committed:
Cleanup tests

kevinglover committed:
Updated style of down page

kevinglover committed:
Updated manager and tests to account for disabled periods

- Banners will not be active if their associated periods are not set to enabled
- Added tests for this behavior
- Cleanup some whitespace

rlskoeser commented on README.rst:
```diff
@@ -8,7 +8,10 @@ Use and Installation:
 * pip install via github url
-* Add ``eultheme`` and ``widget_tweaks`` to **INSTALLED_APPS**
+* Add ``eultheme``, ``widget_tweaks``, and ``downtime`` to **INSTALLED_APPS**
+* Add ``eultheme.middleware.DownpageMiddleware`` to **MIDDLEWARE_CLASSES** after
```
Occurs to me, perhaps we should add sample code for this so it is easier to cut & paste into settings?

rlskoeser commented on eultheme/management/commands/end_downtime.py:
```diff
@@ -0,0 +1,34 @@
+# Based on django-downtime downtime_end command
```
Is it possible to keep the same command name, so that nothing needs to change when this gets fixed in django-downtime?  I don't know how django decides which one to use (order in installed apps?), but if it's something we can control that seems like something we could document.

rlskoeser commented on eultheme/managers.py:
```diff
@@ -0,0 +1,44 @@
+    def active(self):
+        return self.filter(disabled=False, period__enabled=True)
+
+    def first(self):
```
Django queryset already provides first - https://docs.djangoproject.com/en/1.8/ref/models/querysets/#first

kevinglover replied:
But that was introduced in Django 1.6. For 1.5 support, it's needed. I can add a note.

rlskoeser replied:
oh! fair enough.  Yes, please add a comment that it's needed to support django 1.5

rlskoeser commented on eultheme/managers.py:
```diff
@@ -0,0 +1,44 @@
+
+class BannerQuerySet(models.query.QuerySet):
+
+    def active(self):
```
A brief comment/description here would be helpful.  (i.e., why is active more than just not disabled?)

rlskoeser commented on eultheme/managers.py:
```diff
@@ -0,0 +1,44 @@
+
+        return self
+
+    def deployed(self):
```
brief comment/description here would be helpful

rlskoeser commented on eultheme/managers.py:
```diff
@@ -0,0 +1,44 @@
+            now = datetime.datetime.utcnow().replace(tzinfo=utc)
+        else:
+            now = datetime.datetime.now()
+        query_show = Q(show_on_date__lte=now) & (Q(period__end_time__gte=now) | Q(period__end_time=None))
```
Are you sure this is combining the way you want it to? And even if it is, I wonder if it would be more readable to use parentheses and/or line breaks to make it more obvious when reading the code what we are filtering on.

Also, I wonder if would be better to use isnull instead of =None - https://docs.djangoproject.com/en/1.8/ref/models/querysets/#std:fieldlookup-isnull

rlskoeser commented on eultheme/managers.py:
```diff
@@ -0,0 +1,44 @@
+
+        indefinite = banners.filter(period__end_time=None)
+
+        if indefinite:
```
Why are indefinite banners returned first/separately?

rlskoeser commented on eultheme/middleware.py:
```diff
@@ -0,0 +1,23 @@
+class DownpageMiddleware(DowntimeMiddleware):
+
+    def process_request(self, request):
+        def get_client_ip(request):
```
I think it would make more sense (and be more readable) to define this as a class method rather than redefining every time we process a request.  I _think_ it would be more efficient too, but I don't actually know.

rlskoeser commented on eultheme/middleware.py:
```diff
@@ -0,0 +1,23 @@
+
+    def process_request(self, request):
+        def get_client_ip(request):
+            x_forwarded_for = request.META.get('HTTP_X_FORWARDED_FOR')
```
My understanding is that HTTP_X_FORWARDED_FOR is easier to fake, so I wonder about not using this one if we don't need it.  If we do decide to keep it (probably not such a huge risk in this case), please document the splitting you're doing (I'd forgotten - or maybe never knew - that it could have multiple hosts).

kevinglover replied:
Proxies sometimes return multiple IPs in the HTTP_X_FORWARDED_FOR header

rlskoeser commented on eultheme/middleware.py:
```diff
@@ -0,0 +1,23 @@
+        downtime_request = super(DownpageMiddleware,self).process_request(request)
+
+        if downtime_request and \
+            settings.DOWNTIME_ALLOWED_IPS and \
```
Recommend using `getattr(settings, 'DOWNTIME_ALLOWED_IPS', [])` (or something like that) here so that the middleware doesn't break if no IPs are configured.

rlskoeser commented on eultheme/models.py:
```diff
@@ -1,3 +1,58 @@
+        super(Banner, self).save(*args, **kwargs)
+
+    @property
+    def period_has_started(self):
```
docstrings for these properties would be helpful

rlskoeser commented on eultheme/models.py:
```diff
@@ -1,3 +1,58 @@
+            return {'indefinite':True}
+
+    @classmethod
+    def do_observe_period_saved(cls, sender, instance, created, **kwargs):
```
I'm unclear on what this is for / doing

rlskoeser commented on eultheme/templates/eultheme/site_base.html:
```diff
@@ -42,13 +42,16 @@
-          <button class="btn btn-default">OK</button>
-        </div>
-      {% endif %}
+    {% if ENABLE_BETA_WARNING %}
```
Not related to your current changes, but I wonder if this should be a snippet also?

rlskoeser commented on eultheme/templates/downtime/downtime.html:
```diff
@@ -0,0 +1,269 @@
+
```
Would it be very hard to add block tags to make this extendable if projects wanted to customize?

rlskoeser commented on eultheme/tests.py:
```diff
@@ -1,3 +1,77 @@
+            days=0
+        )
+
+        self.active_banner = Banner.objects.get_deployed().first()
```
It's hard for me to tell if you're hitting all the cases here.  Might be worth adding comments and doing them systematically? past/present/future with combinations of active and no end?

rlskoeser commented on eultheme/tests.py:
```diff
@@ -1,3 +1,77 @@
+        # Set banner to disabled
+        self.b1.disabled = True
+        self.b1.save()
 
```
It would probably be a good idea to add tests for the custom queryset filters too.

kevinglover committed:
Updated tests

kevinglover committed:
Add comments to codebase

kevinglover committed:
Refactored client ip method to class method

- Cleaned up middleware get_client_ip method
- Renamed method to client_ip

kevinglover committed:
Cleaned up context processor

- Combined return dictionary
- Added check to banner display to not display if the site is showing the downtime page

kevinglover committed:
Refactored downtime page to use site_base

- Extended site_base for better customization
- Removed whitespace

kevinglover committed:
Wrapped banners in blocks for greater customization

kevinglover committed:
Fixed issue with middleware

- Corrected class method

kevinglover committed:
Update to reflect updated downtime

- Changed references to eultheme.DowntimePeriod to downtime.Period
- Updated tests and migrations
- Changed Banner verbose name to Downtime Banner for clarity

kevinglover committed:
Removed end_downtime command

- End downtime handled by django-downtime

kevinglover committed:
Added disable_downtime command

- New command to disable all enabled downtime periods.
- Helpful for bringing a site out of scheduled maintenance mode.

kevinglover committed:
Cleaned up downtime property in models

- Updated to return duration dict

kevinglover committed:
Added travis ci script

kevinglover committed:
Corrected travis file name

kevinglover committed:
Update travis path

kevinglover committed:
Correct pep8 inconsistencies

- Minor python syntax changes

kevinglover committed:
Exclude django 1.7+ from python 2.6

- Django dropped support for python 2.6 at 1.7

kevinglover committed:
Fix indentions in travis file

kevinglover committed:
Fix for travis ci

kevinglover committed:
Add viewport meta tag

kevinglover committed:
Fixed site_name meta property

- Removed reference to Readux and added general statement
